components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -182,12 +196,71 @@ abstract class AbstractFetchDownloadService : Service() {
             startDownloadJob(download.id)
         }
 
+        notificationUpdateScope.launch {
+            while (isActive) {
+                delay(PROGRESS_UPDATE_INTERVAL)
+                updateDownloadNotification()
+            }
+        }
+
         return super.onStartCommand(intent, flags, startId)
     }
 
+    /***
+     * Android rate limits notifications being sent, so we must send them on a delay so that
+     * notifications are not dropped
+     */
+    private fun updateDownloadNotification() {
+        synchronized(context) {
+            for (download in downloadJobs.values) {
+                // Dispatch the corresponding notification based on the current status
+                val notification = when (download.status) {
+                    DownloadJobStatus.ACTIVE -> {
+                        synchronized(context) {
+                            DownloadNotification.createOngoingDownloadNotification(
+                                context,
+                                download.state,
+                                download.currentBytesCopied
+                            )
+                        }
+                    }
+
+                    DownloadJobStatus.PAUSED -> {
+                        synchronized(context) {
+                            DownloadNotification.createPausedDownloadNotification(context, download.state)
+                        }
+                    }
+
+                    DownloadJobStatus.COMPLETED -> {
+                        synchronized(context) {
+                            DownloadNotification.createDownloadCompletedNotification(context, download.state)
+                        }
+                    }
+
+                    DownloadJobStatus.FAILED -> {
+                        DownloadNotification.createDownloadFailedNotification(context, download.state)
+                    }
+
+                    DownloadJobStatus.CANCELLED -> { return }
+                }
+
+                synchronized(context) {
```

Comment:
Here as well. This is already in the outter synchronized block.

Reply:
Fixed :)